tests/unit/defrag.cpp

```diff
@@ -0,0 +1,522 @@
+		     struct eblob_ram_control *rctl __attribute_unused__,
+		     int fd,
+		     uint64_t data_offset,
+		     void *priv,
+		     void *thread_priv __attribute_unused__)
+{
+	int ret = 0;
+	if (dc->flags & BLOB_DISK_CTL_REMOVE)
+		err(EX_SOFTWARE, "removed dc occured");
+
+	iterator_private &ipriv = *static_cast<iterator_private*>(priv);
+	auto &items = ipriv.items;
+	if (ipriv.number_checked >= items.size())
+		err(EX_SOFTWARE, "index out of range. index: %zd, size: %zd", ipriv.number_checked, items.size());
+
+	auto &item = items[ipriv.number_checked];
+	if (item.removed)
+		err(EX_SOFTWARE, "item removed: %s", eblob_dump_id(item.hashed_key.id));
+	if (item.checked)
+		errx(EX_SOFTWARE, "item already checked. %s", eblob_dump_id(item.hashed_key.id));
+	if (dc->data_size != item.value.size())
+		err(EX_SOFTWARE, "sizes mismatch for item[%zd]=%s, sizes: %zd, %zd", ipriv.number_checked, eblob_dump_id(item.hashed_key.id), dc->data_size, item.value.size());
+
+	std::vector<char> data(dc->data_size);
+	if ((ret = __eblob_read_ll(fd, data.data(), dc->data_size, data_offset)))
+		err(EX_SOFTWARE, "can't read data from fd: %d, key: %s, strerror: %s", fd, eblob_dump_id(item.hashed_key.id), strerror(ret));
+
+	if (data != item.value)
+		err(EX_SOFTWARE, "content of value differ"); // TODO: улучшить все сообщения
+
+	item.checked = true;
+	++ipriv.number_checked;
+	return 0;
+}
+
+
+int datasort(eblob_wrapper &wrapper, std::vector<size_t> indices) {
+	size_t number_bases = indices.size();
+	assert(std::is_sorted(indices.begin(), indices.end(), std::less));  // Note: less functor guarantees uniqeness of indices
+	assert(indices.back() < number_bases);
+
+	datasort_cfg dcfg;
+	std::vector<eblob_base_ctl*> bctls(number_bases);
+	eblob_base_ctl *bctl;
+	size_t index = 0;
+	size_t loop_index = 0;
+	list_for_each_entry(bctl, &wrapper.get()->bases, base_entry) {
+		if (loop_index == indices[index]) {
+			bctls[index] = bctl;
+			++index;
+		}
+
+		++loop_index;
+		if (index == number_bases)
+			break;
+	}
+
+	init_datasort_cfg(wrapper, dcfg);
+	dcfg.bctl = bctls.data();
+	dcfg.bctl_cnt = bctls.size();
+	return eblob_generate_sorted_data(&dcfg);
+}
+
+
+/**
+ * 1) Make two bases with 100 records each.
+ *    State: data.0(unsorted, 100 records), data.1(unsorted, 100 records)
+ * 2) Defrag first base
+ *    State: data.0(sorted, 100 records), data.1(unsorted, 100 records)
+ * 3) Remove half of first base.
+ *    State: data.0(sorted, 50 records), data.1(unsorted, 100 records)
+ * 4) Defrag first base.
+ *    State: data.0(sorted with view, 50 records), data.1(unsorted, 100 records)
+ * 5) Check that bases contains all 150 records
+ */
+BOOST_AUTO_TEST_CASE( test_first_base_sorted_second_base_unsorted ) {
+	const size_t RECORDS_IN_BLOB = 100;
+	const size_t TOTAL_RECORDS = 2 * RECORDS_IN_BLOB;
+	const size_t RECORDS_TO_REMOVE = RECORDS_IN_BLOB / 2;
+	std::mt19937 gen(42);
+	std::uniform_int_distribution<unsigned> dist;
+	auto gen_random_data = [&gen, &dist]() {
```

Comment:
This lambda seems superfluous.

Reply:
Ok. What do you think about that?
https://github.com/interiorem/eblob/pull/36/commits/31564d4577e391856ad10f8b537eeb3fbc03961a

Reply:
At first comment I commented about why to write lambda that simplifies nothing. You could use `generate_random_data(gen, dist)` instead of `gen_random_data()`. Where is profit, economy of 10 symbols?

The commit that you made doesn't relate to the question. Note that there used wrong code style.

Reply:
I just don't like keep in mind that I have separate objects like gen and dist. It's cleaner for me if I could use one class for that work.